tls/doc/tls_introduction.md

```diff
@@ -0,0 +1,263 @@
+
+Think about it:
+
+* How does the company providing hundreds of thousands, or even millions, of certificates plan to reliably establish the identity of a certificate holder?
+* How does the CA check the services are legitimate?
+* If the issue holder’s services are now legitimate, how can the CA tell this will still be the case in a week or month?
+* What if the issue holder uses the certificate to host someone else’s malicious service?
+* What if a legit service was hijacked by an attacker?
+
+What the CAs really verify is that someone paid for their certificates. Preferably, with a credit card. While that might help law enforcement agencies to find the person behind a scam, it does nothing to protect you when you establish a TLS connection. You might find it consoling that some responsible people are put behind bars, but your data and potentially money are already gone by that time.
+
+A TLS server certificate tells you whether the hostname of the server matches the one in the certificate. Period.
+
+That’s its purpose, that’s what it does well, and it is also the only thing it does.
+
+If you understood that fact, you will also understand why certificates issued by a non-profit organization like “Let’s Encrypt” are exactly as good or bad as one issued by Verisign.
+
+### Client Certificates
+
+One more part of authentication that we haven’t touched yet is authentication in the opposite direction. The reason for this is that in almost all cases servers don’t use TLS to establish the client's authenticity but mechanisms on the application layer like a login and a session ID or tokens.
+
+TLS supports client certificates, but they are seldom used. Mostly because of the [too-many-keys problem](#the-too-many-keys-problem) we discussed earlier. Services typically do not trust public CAs enough to accept client certificates issued by someone outside their own organization.
+
+That being said, there are cases where client certificates are indeed useful. Mainly in machine-to-machine communication. IT departments often use their own CAs to issue client certificates that they hand out to developers of applications that access company services via TLS. The application then has to present its client certificate during the handshake phase. Only if the service can successfully authenticate the client via its client certificate, will it accept the TLS connection.
+
+### Truststores and Keystores
+
+The words "keystore" and "truststore" are often used as synonyms. This is not precise. Truststores are a specific subset of keystores containing someone else’s public keys or certificates. If you install root CA certificates on a machine, they go into a truststore.
+
+In the Java world the distinction is very strict. When the official documentation says "keystore," it really only refers to the place where you store your own keys.
```

Comment:
ok to keep, i think but wanted to ping as only language-specific use case so far